File: app/src/main/java/ru/roscha_akademii/medialib/update/UpdateService.java
package ru.roscha_akademii.medialib.update;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.util.Log;

import com.pushtorefresh.storio.sqlite.StorIOSQLite;

import java.util.ArrayList;

import javax.inject.Inject;
import javax.inject.Named;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ru.roscha_akademii.medialib.common.MediaLibApplication;
import ru.roscha_akademii.medialib.net.VideoApi;
import ru.roscha_akademii.medialib.net.model.Video;
import ru.roscha_akademii.medialib.net.model.VideoAnswer;

import static ru.roscha_akademii.medialib.video.VideoDbModule.VIDEO_DB;

public class UpdateService extends Service {
    /**
     * Intent для запуска этого сервиса
     *
     * @param context - any context
     * @return intent
     */
    public static Intent getStartIntent(Context context) {
        return new Intent(context, UpdateService.class);
    }

    @Inject
    UpdateScheduler updateScheduler;

    @Inject
    VideoApi api;

    @Inject
    @Named(VIDEO_DB)
    StorIOSQLite videoDb;

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d("happy", "UpdateService onCreate");

        ((MediaLibApplication) getApplicationContext()).component().inject(this);

        update();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d("happy", "UpdateService onDestroy");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        return START_NOT_STICKY;
    }

    void update() {
        Call<VideoAnswer> call = api.getVideoList();
        call.enqueue(new Callback<VideoAnswer>() {
            @Override
            public void onResponse(Call<VideoAnswer> call, Response<VideoAnswer> response) {
                try {
                    VideoAnswer answer = response.body();
                    Log.d("happy", "got videos " + answer.list.size());

                    saveVideos(answer.list);

                    updateScheduler.updateCompleted();
                } catch (Exception ignore) {
                    // ничего, в следующий раз получится
                }
                stopSelf();
            }

            @Override
            public void onFailure(Call<VideoAnswer> call, Throwable t) {
                stopSelf();
            }
        });
    }

    void saveVideos(ArrayList<Video> list) {
        videoDb
                .put()
                .objects(list)
                .prepare()
                .executeAsBlocking();

    }
}

File: app/src/main/java/ru/roscha_akademii/medialib/update/UpdateService.kt
package ru.roscha_akademii.medialib.update

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.util.Log
import org.greenrobot.eventbus.EventBus
import ru.roscha_akademii.medialib.book.model.remote.BookUpdate
import ru.roscha_akademii.medialib.common.MediaLibApplication
import ru.roscha_akademii.medialib.update.event.DataDownloaded
import ru.roscha_akademii.medialib.video.model.remote.VideoUpdate
import javax.inject.Inject

class UpdateService : Service() {
    @Inject
    lateinit var videoUpdater: VideoUpdate

    @Inject
    lateinit var bookUpdater: BookUpdate

    @Inject
    lateinit var updateScheduler: UpdateScheduler

    @Inject
    lateinit var bus: EventBus

    val CHANNELS = 2
    var inProgress: Int = 0
    var success: Int = 0

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    val updateCallback = object : UpdateCallback {
        override fun onSuccess() {
            success++
            bus.post(DataDownloaded())
            somethingFinished()
        }

        override fun onFail() {
            somethingFinished()
        }

    }

    fun somethingFinished() {
        inProgress--

        if (inProgress == 0) {
            if (success == CHANNELS) {
                updateScheduler.updateCompleted()
            }

            stopSelf()
        }
    }

    fun startUpdate() {
        inProgress = CHANNELS
        success = 0

        videoUpdater.update(updateCallback)
        bookUpdater.update(updateCallback)
    }

    override fun onCreate() {
        super.onCreate()
        Log.d("happy", "UpdateService onCreate")

        (applicationContext as MediaLibApplication).component.inject(this)

        startUpdate()

    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("happy", "UpdateService onDestroy")
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {

        return Service.START_NOT_STICKY
    }


    companion object {
        /**
         * Intent для запуска этого сервиса

         * @param context - any context
         * *
         * @return intent
         */
        fun getStartIntent(context: Context): Intent {
            return Intent(context, UpdateService::class.java)
        }
    }
}
